File: neogrid-ndap/src/engine.js
const path = require('path');
const fs = require('fs');
const { hashData } = require('./utils');
const { BinarySMT } = require('./smt');
const { LinkedMMR } = require('./mmr');
const { validateTxStructure, canonicalize, txHash } = require('./tx');
const { uploadToIPFS, canonicalHash } = require('./ipfs');

const RUST_BINARY = path.join(__dirname, '../core-rust/target/release/ndap-engine');

class Engine {
  constructor() {
    this._smt = new BinarySMT();
    this._mmr = new LinkedMMR();
    this._usedNonces = new Set();
    this._ownership = {};
    this._balances = {};
    this._assets = {};
    this._rustAvailable = fs.existsSync(RUST_BINARY);
  }

  rustAvailable() {
    return this._rustAvailable;
  }

  validate_transaction(tx) {
    const structural = validateTxStructure(tx);
    if (!structural.valid) return { valid: false, errors: structural.errors };

    const errors = [];

    if (this._usedNonces.has(tx.nonce)) {
      errors.push(`Replay attack detected: nonce "${tx.nonce}" already consumed`);
    }

    const currentOwner = this._ownership[tx.assetId];
    if (currentOwner && currentOwner !== tx.fromDID) {
      errors.push(`Ownership violation: asset "${tx.assetId}" is owned by "${currentOwner}", not "${tx.fromDID}"`);
    }

    if (tx.amount !== undefined && tx.amount > 0) {
      const balance = this._balances[tx.fromDID] || 0;
      if (balance < tx.amount) {
        errors.push(`Insufficient balance: have ${balance}, need ${tx.amount}`);
      }
    }

    return { valid: errors.length === 0, errors };
  }

  apply_transaction(tx) {
    const validation = this.validate_transaction(tx);
    if (!validation.valid) {
      throw new Error(`Transaction rejected: ${validation.errors.join('; ')}`);
    }

    const prevRoot = this.compute_state_root();
    const hash = txHash(tx);

    this._usedNonces.add(tx.nonce);

    if (tx.assetId && tx.toDID) {
      this._ownership[tx.assetId] = tx.toDID;
      this._smt.set(tx.assetId, tx.toDID);
    }

    if (tx.amount && tx.amount > 0) {
      this._balances[tx.fromDID] = (this._balances[tx.fromDID] || 0) - tx.amount;
      this._balances[tx.toDID] = (this._balances[tx.toDID] || 0) + tx.amount;
    }

    const newRoot = this.compute_state_root();
    const logEntry = this._mmr.appendLinked(hash, prevRoot, newRoot);

    return { prevRoot, newRoot, txHash: hash, logEntry };
  }

  async register_asset(assetId, owner, data) {
    const { cid, contentHash } = await uploadToIPFS(data);
    const prevRoot = this.compute_state_root();

    this._assets[assetId] = { cid, contentHash, owner, data };
    this._ownership[assetId] = owner;
    this._smt.set(assetId, cid);

    const newRoot = this.compute_state_root();
    const entryHash = hashData(`register:${assetId}:${cid}:${owner}`);
    const logEntry = this._mmr.appendLinked(entryHash, prevRoot, newRoot);

    return { assetId, cid, contentHash, root: newRoot, logEntry };
  }

  compute_state_root() {
    return this._smt.computeRoot();
  }

  get_proof(key) {
    return this._smt.getProof(key);
  }

  verify_proof(proof) {
    return this._smt.verifyProof(proof);
  }

  get_log() {
    return this._mmr.getAll();
  }

  get_mmr_root() {
    return this._mmr.currentRoot();
  }

  verify_log_integrity() {
    return this._mmr.verify();
  }

  get_ownership(assetId) {
    return this._ownership[assetId] || null;
  }

  get_balance(did) {
    return this._balances[did] || 0;
  }

  snapshot() {
    return {
      stateRoot: this.compute_state_root(),
      mmrRoot: this.get_mmr_root(),
      logSize: this._mmr.size(),
      usedNonces: this._usedNonces.size,
      assets: Object.keys(this._ownership).length,
      rustEngine: this._rustAvailable,
    };
  }
}

module.exports = { Engine };

File: neogrid-ndap/core-rust/src/engine.rs
use crate::hash::sha256_hex;
use crate::tx::Transaction;
use crate::state::State;
use crate::smt::InclusionProof;
use crate::mmr::LogEntry;

#[derive(Debug)]
pub struct StateTransition {
    pub prev_root: String,
    pub new_root: String,
    pub tx_hash: String,
    pub log_entry: LogEntry,
}

#[derive(Debug)]
pub struct ValidationResult {
    pub valid: bool,
    pub errors: Vec<String>,
}

pub struct Engine {
    state: State,
}

impl Engine {
    pub fn new() -> Self {
        Engine { state: State::new() }
    }

    pub fn validate_transaction(&self, tx: &Transaction) -> ValidationResult {
        let mut errors = Vec::new();

        if tx.from_did.is_empty() { errors.push("Missing: fromDID".to_string()); }
        if tx.to_did.is_empty() { errors.push("Missing: toDID".to_string()); }
        if tx.asset_id.is_empty() { errors.push("Missing: assetId".to_string()); }
        if tx.nonce.is_empty() { errors.push("Missing: nonce".to_string()); }
        if tx.timestamp == 0 { errors.push("Missing: timestamp".to_string()); }
        if tx.from_did == tx.to_did { errors.push("fromDID and toDID must differ".to_string()); }

        if self.state.used_nonces.contains(&tx.nonce) {
            errors.push(format!("Replay attack: nonce \"{}\" already consumed", tx.nonce));
        }

        if let Some(owner) = self.state.ownership.get(&tx.asset_id) {
            if owner != &tx.from_did {
                errors.push(format!("Ownership violation: asset owned by \"{}\"", owner));
            }
        }

        if tx.amount > 0 {
            let balance = self.state.balances.get(&tx.from_did).copied().unwrap_or(0);
            if balance < tx.amount {
                errors.push(format!("Insufficient balance: have {}, need {}", balance, tx.amount));
            }
        }

        ValidationResult { valid: errors.is_empty(), errors }
    }

    pub fn apply_transaction(&mut self, tx: &Transaction) -> Result<StateTransition, String> {
        let result = self.validate_transaction(tx);
        if !result.valid {
            return Err(format!("Transaction rejected: {}", result.errors.join("; ")));
        }

        let prev_root = self.state.state_root();
        let tx_hash = tx.fingerprint();

        self.state.used_nonces.insert(tx.nonce.clone());
        self.state.ownership.insert(tx.asset_id.clone(), tx.to_did.clone());
        self.state.smt.set(&tx.asset_id, &tx.to_did);

        if tx.amount > 0 {
            *self.state.balances.entry(tx.from_did.clone()).or_insert(0) -= tx.amount;
            *self.state.balances.entry(tx.to_did.clone()).or_insert(0) += tx.amount;
        }

        let new_root = self.state.state_root();
        let timestamp = tx.timestamp;
        let log_entry = self.state.mmr.append_linked(&tx_hash, &prev_root, &new_root, timestamp);

        Ok(StateTransition { prev_root, new_root, tx_hash, log_entry })
    }

    pub fn compute_state_root(&self) -> String {
        self.state.state_root()
    }

    pub fn get_proof(&self, key: &str) -> InclusionProof {
        self.state.smt.get_proof(key)
    }

    pub fn snapshot(&self) -> crate::state::StateSnapshot {
        self.state.snapshot()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn make_tx(asset: &str, from: &str, to: &str, nonce: &str) -> Transaction {
        Transaction {
            id: String::new(),
            from_did: from.to_string(),
            to_did: to.to_string(),
            asset_id: asset.to_string(),
            amount: 0,
            nonce: nonce.to_string(),
            timestamp: 1700000000,
            signature: None,
        }
    }

    #[test]
    fn test_apply_and_state_root_changes() {
        let mut engine = Engine::new();
        let r0 = engine.compute_state_root();
        let tx = make_tx("asset-1", "alice", "bob", "nonce-1");
        engine.apply_transaction(&tx).unwrap();
        let r1 = engine.compute_state_root();
        assert_ne!(r0, r1);
    }

    #[test]
    fn test_nonce_replay_rejected() {
        let mut engine = Engine::new();
        let tx = make_tx("asset-1", "alice", "bob", "nonce-dup");
        engine.apply_transaction(&tx).unwrap();
        let tx2 = make_tx("asset-1", "bob", "carol", "nonce-dup");
        let result = engine.apply_transaction(&tx2);
        assert!(result.is_err());
    }

    #[test]
    fn test_ownership_violation_rejected() {
        let mut engine = Engine::new();
        let tx1 = make_tx("asset-X", "alice", "bob", "n1");
        engine.apply_transaction(&tx1).unwrap();
        let tx2 = make_tx("asset-X", "charlie", "dave", "n2");
        let result = engine.apply_transaction(&tx2);
        assert!(result.is_err());
    }
}
